core/services/run_queue.go

```diff
@@ -34,81 +36,188 @@ type RunQueue interface {
+					rq.workers[*run.JobSpecID] = worker
+					atomic.AddInt32(&rq.numWorkers, 1)
+					go rq.runSingleJobSpecWorker(*run.JobSpecID, worker)
+					worker.chAddRun <- run
+				}
+			}
 
-	rq.workersWg.Add(1)
-	go func() {
-		for {
-			rq.workersMutex.Lock()
-			queueCount := rq.workers[runID]
-			if queueCount <= 0 {
-				delete(rq.workers, runID)
-				numberRunQueueWorkers.Set(float64(len(rq.workers)))
-				rq.workersMutex.Unlock()
-				break
+		case jobSpecID := <-rq.chWorkerDone:
+			delete(rq.workers, jobSpecID)
+			if n := atomic.AddInt32(&rq.numWorkers, -1); n < 0 {
+				panic("numWorkers should never be < 0")
 			}
-			rq.workers[runID] = queueCount - 1
-			rq.workersMutex.Unlock()
 
-			if err := rq.runExecutor.Execute(run.ID); err != nil {
-				logger.Errorw(fmt.Sprint("Error executing run ", runID), "error", err)
+		case <-rq.chStop:
+			for _, worker := range rq.workers {
+				worker.chStop <- struct{}{}
+				<-worker.chDone
+				if n := atomic.AddInt32(&rq.numWorkers, -1); n < 0 {
+					panic("numWorkers should never be < 0")
+				}
 			}
+			return
 		}
+	}
+}
 
-		rq.workersWg.Done()
+func (rq *runQueue) runSingleJobSpecWorker(jobSpecID models.ID, worker *singleJobSpecWorker) {
+	defer close(worker.chDone)
+
+	promNumberRunQueueWorkersStarted.Inc()
+
+	var (
+		startOnce       sync.Once
+		chWorkerStarted = make(chan struct{})
+		chResume        = make(chan []models.ID)
+		chWorkComplete  = make(chan struct{})
+	)
+
+	// The worker goroutine accepts job run requests from the coordinator loop below.  It
+	// can accept further requests after work has begun.  Once its queue of work is empty,
+	// it sends a "work finished" message to that loop, which may or may not cause it to
+	// spin down, depending on whether further requests are already enqueued.
+	go func() {
+		for runsRequested := range chResume {
+			// We have to wait until the worker has actually received a job before indicating
+			// to the coordinator loop that work has started.  If we do this before the worker's
+			// for loop starts, there's a race that can allow the coordinator to receive a chStop
+			// message and shut down the worker even though runs are queued.
+			startOnce.Do(func() { close(chWorkerStarted) })
+
+			for _, runID := range runsRequested {
+				select {
+				case <-worker.chStop:
+					return
+				default:
+				}
+
+				promNumberRunsQueued.Inc()
+
+				if err := rq.runExecutor.Execute(&runID); err != nil {
+					logger.Errorw(fmt.Sprint("Error executing run ", runID.String()), "error", err)
+				}
+			}
+			select {
+			case chWorkComplete <- struct{}{}:
+			case <-worker.chStop:
+				return
+			}
+		}
 	}()
+
+	// The coordinator loop accepts "do job run" requests from orchestrateWorkers, "work finished"
+	// signals from the worker goroutine, and "stop" messages from the Chainlink application,
+	// coordinating them such that we can avoid races when messages arrive simultaneously.
+	var runsRequested []models.ID
+	for {
+		select {
```

Comment:
Again, the cognitive complexity of this select statement is high, could we break it down into some self describing code?
```
case run := <-worker.chAddRun:
  rq.ResumeRun(run)
case <-chWorkComplete:
  rq.ResumeOrShutdown()
case <-worker.chStop: 
  rq.Shutdown()
```